docs/markdown/Reference-manual.md

```diff
@@ -600,8 +600,12 @@ be passed to [shared and static libraries](#library).
   depends on such as a symbol visibility map. The purpose is to
   automatically trigger a re-link (but not a re-compile) of the target
   when this file changes.
-- `link_language` since 0.51.0 makes the linker for this target
-  be for the specified language. This is helpful for multi-language targets.
+- `link_language` since 0.51.0 (broken until 0.55.0) makes the linker for this
+  target be for the specified language. It is generally unnecessary to set
+  this, as meson will detect the right linker to use in most cases. There are
+  only two cases where this is needed. One, your main function in an
+  executable is not in the language meson picked, or second you want to force
+  a library to expose only one ABI, such as using C++ with `extern "C"`.
```

Comment:
I think this comment for C++ is misleading. This option does not help for this, `extern "C"` is enough and adding `link_language` won't forbid the library to export also C++ symbols. What will prevent is to import symbols from standard C++ libraries. In different cases you want to use C++ as a C with additional features but for dependencies or other reasons you don't want to use more than that.

The `extern "C"` in C++ allows to: 

1. import C symbols as C symbols (so for instance you can use `printf` from C++)
2. export C++ symbols as C symbols (so for instance you can use a `my_printf` written in C++ from a C program)

Reply:
The C linker will die a painful death if you have any C++ symbols exposed, which is something a number of people have wanted: write in C++ but ensure only C symbols are exposed by linking with the C linker.

Reply:
Maybe I don't follow. I tried with your branch this:

- meson.build
```python
project('foo', 'cpp', 'c')

shared_library('foo', ['main.cpp'], link_language : 'c')
```
- main.cpp
```c++
void foo(int) { }
```
It compiles, the shared object file contains C++ symbols `objdump -TC libfoo.so` gives:
```

libfoo.so:     file format elf64-x86-64

DYNAMIC SYMBOL TABLE:
0000000000000000  w   D  *UND*  0000000000000000 __cxa_finalize
0000000000000000  w   D  *UND*  0000000000000000 _ITM_registerTMCloneTable
0000000000000000  w   D  *UND*  0000000000000000 _ITM_deregisterTMCloneTable
0000000000000000  w   D  *UND*  0000000000000000 __gmon_start__
00000000000010f9 g    DF .text  000000000000000a foo(int)
```
`ninja` is using the wanted linker:
```
$ ninja -v
[1/2] ccache c++ -Ifoo@sha -I. -I.. -I/home/user/install/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -g -fPIC -MD -MQ 'foo@sha/main.cpp.o' -MF 'foo@sha/main.cpp.o.d' -o 'foo@sha/main.cpp.o' -c ../main.cpp
[2/2] cc  -o libfoo.so 'foo@sha/main.cpp.o' -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,libfoo.so -Wl,--end-group
```

Reply:
Consider this (assume there's a match .h):

```cpp
extern "C" int foo() { return 1; }
```

This only exposes C linkage symbols, you can link this with the C linker and everything is good.

But this:

```cpp
#include <string>

std::string bar() { return std::string{"foo"}; }
extern "C" int foo() { return 1; }
```

Will not link with the c linker. This is the case I'm trying to describe.

Reply:
Yes, I perfectly understood. But the linker is not failing because the module is exporting `bar` but because is importing `std::string::string`. So maybe just changing `a library to expose only one ABI` to `a library to import only one ABI` in your description would be enough.